### PR TITLE
update to SDKv6 + config panel

### PR DESCRIPTION
**Pull request by szogoon**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,57 +1,411 @@
...
+  switch (tabId) {
+    case "api-key-form":
+      tooltip += "<span>TIP:</span> Remember to check if your API Key is valid for Geofencing API. If your are unsure, consult the <a href='https://developer.tomtom.com/geofencing-api/tutorials/fence-creation' target='_blank'>Tutorial</a>.";
+      break;
+    case "admin-key-form":
+      tooltip += "Check <i>project ID</i> details in <a href='https://developer.tomtom.com/geofencing-api/geofencing-api-documentation-configuration-service/register-admin-key' target='_blank'>geofencing documentation</a>";
+      break;
+    case "project-id-form":
+      tooltip += "Check <i>admin</i> key details in <a href='https://developer.tomtom.com/geofencing-api/geofencing-api-documentation-projects-service/add-new-project' target='_blank'>geofencing documentation</a>";
+      break;
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js:55]: I think it would be safer to add a default case as well.

File: app.js [hunk 2]
```diff
@@ -1,57 +1,411 @@
...
+  }
+  document.getElementById("tooltips").innerHTML = tooltip;
+  document.getElementById(currentTab).style.display = "none";
+  document.getElementById(tabId).style.display = "block";
+  var icons = document.getElementsByClassName("progress-icon");
+  icons.forEach(function(icon) {
+    icon.style.background = "#8dc3eb";
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js]: Store the colors in variables.

File: app.js [hunk 3]
```diff
@@ -1,57 +1,411 @@
...
+  var selectedIcon = document.querySelector("div[for=" + tabId + "]");
+  selectedIcon.style.background = "#fff";
+  selectedIcon.style.color = "#8dc3eb";
+  currentTab = tabId;
+}
+
+function validateForm() {
+  var currentTabElem = document.getElementById(currentTab);
+  var fields = currentTabElem.getElementsByTagName("input");
+  var i, valid = true;
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js]: The `i` variable is `undefined` here and it's a bit confusing. It’s a good practice to put each variable on its own line for improved readability.

File: app.js [hunk 4]
```diff
@@ -1,57 +1,411 @@
...
+  if (document.querySelector("[for='generated-admin-key']").style.display == "block") {
+    valid = true;
+  }
+  for (i = 0; i < fields.length; i++) {
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js]: You could write `for (var i = 0; ...)` instead of declaring the `i` variable in line `65`.

File: app.js [hunk 5]
```diff
@@ -1,57 +1,411 @@
...
+  if (validateForm()) {
+    apiKey = document.getElementById("api-key").value;
+    showTab("admin-key-form");
+  };
+});
+
+document.getElementById("save-admin-key").addEventListener("click", function() {
+  if (validateForm()) {
+    var style = window.getComputedStyle(document.querySelector("[for='generated-admin-key']"));
+    if ( style.getPropertyValue('display') != "none" ) {
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js]: It might be safer to use `!==` to check the value and type equality.

File: app.js [hunk 6]
```diff
@@ -1,57 +1,411 @@
...
+document.getElementById("back-to-api-key").addEventListener("click", function() {
+  showTab("api-key-form");
+});
+
+document.getElementById("back-to-admin-key").addEventListener("click", function() {
+  showTab("admin-key-form");
+});
+
+document.getElementById("provide-admin-key-tab").addEventListener("click", function() {
+  if (document.querySelector("[for='generated-admin-key']").style.display != "block") {
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js]: `!==`

File: app.js [hunk 7]
```diff
@@ -1,57 +1,411 @@
...
+    document.querySelector("[for='secret']").style.display = "block";
+  }
+})
+
+document.getElementById("gen-admin-key").addEventListener("click", function() {
+  var secret = document.getElementById("secret").value;
+  generateAdminKey(secret).then(function(key) {displayAdminKey(key)});
+})
+
+document.getElementById("how-to-get-api-key").addEventListener("click", function () {location.href="https://developer.tomtom.com/how-to-get-tomtom-api-key"});
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js]: The formatting looks a bit off.

File: app.js [hunk 8]
```diff
@@ -1,57 +1,411 @@
...
+      this.geometry.coordinates[1] = [e.lngLat.lng, e.lngLat.lat];
+      var features = turf.featureCollection([
+        turf.point(this.geometry.coordinates[0]),
+        turf.point(this.geometry.coordinates[1])
+      ]);
+      var geoJsonData = turf.envelope(features);
+      this.redraw(geoJsonData);
+    };
+    let onStartDrawing = function(event) {
+      if (drawState != "cancel") {
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js]: `!==`

File: app.js [hunk 9]
```diff
@@ -133,34 +480,42 @@ function displayFence(data) {
...
+        data.geometry = geoJsonData.geometry;
+        return data;
       case "Rectangle":
         geoJsonData = turf.envelope(data.geometry);
-        break;
+        data.geometry = geoJsonData.geometry;
+        return data;
       case "Corridor":
-        geoJsonData = turf.buffer(
+        geoJsonData =  turf.buffer(
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js]: Unnecessary spacing ( ͡° ͜ʖ ͡°)

File: app.js [hunk 10]
```diff
@@ -379,161 +765,75 @@ function processAdditionalDataResponse(additionalDataResponse) {
...
-      map.on("dblclick", function() {
-        self.geometry = convertLineStringToPolygon(self.geometry);
-        self.redraw(self.geometry);
-        self.endDrawing(self.geometry, self.geoJson);
-      });
-    }
-  );
-});
+function clearButtonsState() {
+  document.getElementsByClassName("choice-button").forEach(button => {
```
Comment (BartlomiejCiszewski-TomTom) [inline app.js:829]: Bear in mind that arrow functions won't work in IE11.

File: style.css [hunk 11]
```diff
@@ -1,205 +1,537 @@
...
+  color: inherit;
+}
+
+#config-form {
+  background: #8dc3eb;
+  height: 100vh;
+  min-height: 785px;
+}
+
+#worek {
```
Comment (BartlomiejCiszewski-TomTom) [inline style.css]: Worek? ( ͡◉ ͜ʖ ͡◉)

**Review by BartlomiejCiszewski-TomTom — APPROVED** [review]
Pre-approved

File: style.css (hunk 11 above)
Comment (szogoon) [reply style.css]: renamed to form-container

File: app.js (hunk 10 above)
Comment (szogoon) [reply app.js:829]: Yeah, I'm aware of this and we decided to not support IE in favor of having better readable code.

**Review by BartlomiejCiszewski-TomTom — APPROVED** [review]

File: app.js (hunk 1 above)
Comment (szogoon) [reply app.js:55]: I've made first one a default.

File: app.js [hunk 12]
```diff
@@ -1,57 +1,415 @@
...
-    closeOnMapClick: false,
-    close: null
-  })
-  .addTo(map)
-  .addContent(document.getElementById("inputs"));
+tt.setProductInfo("Fence manager", "2.1");
+var map;
 
 var inputPopup =
   '<div class="form">' +
```
Comment (falseinput) [inline app.js]: Inconsistent quoting in this block. Consider using template literals to increase readability.

File: app.js [hunk 13]
```diff
@@ -1,57 +1,415 @@
...
+
+document.getElementById("how-to-get-api-key").addEventListener("click", function() {
+  location.href="https://developer.tomtom.com/how-to-get-tomtom-api-key"
+});
+
+document.getElementById("config").addEventListener("click", showConfigForm);
+
+function retrieveProjects() {
+  axios
+  .get(
```
Comment (falseinput) [inline app.js]: Consider using template literals

File: app.js [hunk 14]
```diff
@@ -1,57 +1,415 @@
...
+  )
+  .then(function(response) {
+    if (response.data.projects.length > 0) {
+      response.data.projects.forEach(function(project) {
+        addProjectToProjectsList(project);
+      })
+    }
+    else {
+      axios
+      .post(
```
Comment (falseinput) [inline app.js]: Consider using template literals

File: app.js [hunk 15]
```diff
@@ -379,161 +767,75 @@ function processAdditionalDataResponse(additionalDataResponse) {
...
       document.onkeydown = null;
     }
   };
 }
 
-function displayModal(message) {
-  modalContent.innerText = message;
-  modal.style.display = "block";
+function getBounds(geoJson) {
+  let envelope = turf.envelope(geoJson);
```
Comment (falseinput) [inline app.js]: This function returns bounds, so I think using turf.bbox https://turfjs.org/docs/#bbox would be sufficient.

File: app.js [hunk 16]
```diff
@@ -65,39 +423,33 @@ function getFences() {
...
     });
 }
 
-function getFenceDetails(fence) {
-  var counter =
-    arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
+function getFenceDetails(fence, counter = 0) {
   var retryTimes = 5;
   return axios
     .get(geofencingApiURL + "fences/" + fence.id + "?key=" + apiKey)
```
Comment (falseinput) [inline app.js]: Consider using template literals.

File: app.js [hunk 17]
```diff
@@ -1,57 +1,415 @@
...
+      displayFence(fence);
+    })
+    let geoJson = turf.featureCollection(transformedFences);
+    let bounds = getBounds(geoJson);
+    map.fitBounds(bounds, { padding: { top: 15, bottom:15, left: 15, right: 15 }, animate: false });
+  })
+}
+
+function generateAdminKey(secret) {
+  return axios
```
Comment (falseinput) [inline app.js:426]: Consider using template literals.

File: app.js [hunk 18]
```diff
@@ -1,57 +1,415 @@
...
+  });
+  
+  var attributions = [
+    '<a href="https://www.tomtom.com/mapshare/tools/" target="_blank">Report map issue</a>'
+  ];
+  map.getAttributionControl().addAttribution(attributions);
+  
+  map.addControl(new tt.NavigationControl(), "top-left");
+  
+  new Foldable(".js-foldable", "top-right");
```
Comment (falseinput) [inline app.js]: Should it work on mobile? Maybe this whole folding capability is not needed.

File: app.js [hunk 19]
```diff
@@ -321,51 +710,50 @@ function saveFence(fenceData, geoJson) {
...
     .catch(function(err) {
-      displayModal(
-        "There was an error while saving fences: " + err.response.data.message
-      );
+      displayModal("error", saveFenceErrorMsg(err));
+      polygon.remove();
     });
 }
 
 function findGeometry() {
```
Comment (falseinput) [inline app.js]: This functionality works a little bit like magic. When I move a map to the York in UK and type "York" I get polygon of NY. When I type "pizza" I get no feedback (that's less surprising ;) ).
Did you consider using searchbox plugin with location bias to help with selection of geometry?

File: app.js [hunk 20]
```diff
@@ -1,57 +1,415 @@
-var geofencingApiURL = 'https://api.tomtom.com/geofencing/1/';
+var currentTab = "api-key-form";
```
Comment (falseinput) [inline app.js]: Let's stick to using either var or const and let.

**Review by falseinput — CHANGES_REQUESTED** [review]

File: app.js (hunk 15 above)
Comment (szogoon) [reply app.js]: If a user don't have any fences in his project then bbox returns [Infinity, Infinity, -Infinity, -Infinity] which can't be used as bounds. It throws then: Invalid LngLat latitude value: must be between -90 and 90

File: app.js (hunk 18 above)
Comment (szogoon) [reply app.js]: Good point. I'll remove it.

File: app.js (hunk 19 above)
Comment (szogoon) [reply app.js]: implemented Searchbox

File: app.js (hunk 20 above)
Comment (szogoon) [reply app.js]: replaced var with const and let

**Comment by szogoon** [comment]
all changes applied

**Review by falseinput — APPROVED** [review]

**Pull request merged**